src/dataflow/src/source/postgres.rs

```diff
@@ -408,7 +454,52 @@ impl PostgresSourceReader {
                             }
                             self.metrics.lsn.set(self.lsn.into());
                         }
-                        Origin(_) | Relation(_) | Type(_) => {
+                        Relation(relation) => {
+                            let rel_id = relation.rel_id();
+                            if let Some(source_table) =
+                                self.source_tables.get(&rel_id.try_into().unwrap())
+                            {
+                                // Start with the cheapest check first, this will catch the majority of alters
+                                if source_table.columns.len() != relation.columns().len() {
+                                    return Err(Fatal(anyhow!(
+                                        "source table {} with oid {} has been altered",
+                                        source_table.name,
+                                        source_table.relation_id
+                                    )));
+                                }
+                                // todo(nharring): use an array instead of vec to avoid heap allocation
+                                let cols: Vec<PostgresColumn> = relation
+                                    .columns()
+                                    .into_iter()
+                                    .map(|c| PostgresColumn {
+                                        name: c.name().unwrap().to_string(),
+                                        type_oid: c.type_id(),
+                                        type_mod: c.type_modifier(),
+                                        nullable: bool::default(),
+                                        primary_key: bool::default(),
+                                    })
+                                    .collect();
```

Comment:
You can remove `.collect()` from here, remove the `: Vec<PostgresColumn>` from the `let` bindings, and also replace `cols.iter()` with just `cols` below to avoid the vector